File: web/src/lib/alerts.ts
import { api } from "./api";

export type AlertSeverity = "info" | "warning" | "critical";
export type AlertStatus = "open" | "closed";
export type AlertComparator = "gt" | "gte" | "lt" | "lte";

export type Alert = {
  id: string;
  ruleId: string;
  clusterId: string;
  accountName?: string;
  status: AlertStatus;
  severity: AlertSeverity;
  metric: string;
  message: string;
  firingValue: number;
  threshold: number;
  firstSeenAt: string;
  lastSeenAt: string;
  closedAt?: string;
  acknowledgedAt?: string;
  acknowledgedBy?: string;
  ruleName?: string;
};

export type AlertRule = {
  id: string;
  clusterId?: string;
  accountName?: string;
  name: string;
  message: string;
  severity: AlertSeverity;
  metric: string;
  comparator: AlertComparator;
  threshold: number;
  enabled: boolean;
  createdBy?: string;
  createdAt: string;
  updatedAt: string;
};

export type AlertOpenSummary = {
  count: number;
  alerts: Alert[];
};

export type AlertRuleMetrics = {
  metrics: string[];
  comparators: AlertComparator[];
  severities: AlertSeverity[];
};

export async function fetchAlertOpenSummary() {
  return (await api<AlertOpenSummary>("/api/v1/alerts/open-summary")).data;
}

export async function fetchAlerts(params: { status?: AlertStatus; clusterId?: string; limit?: number }) {
  const q = new URLSearchParams();
  if (params.status) q.set("status", params.status);
  if (params.clusterId) q.set("clusterId", params.clusterId);
  if (params.limit) q.set("limit", String(params.limit));
  const suffix = q.toString() ? `?${q}` : "";
  const r = await api<Alert[]>(`/api/v1/alerts${suffix}`);
  return { alerts: r.data ?? [], total: r.meta?.total ?? 0 };
}

export async function acknowledgeAlert(id: string) {
  return (await api<Alert>(`/api/v1/alerts/${encodeURIComponent(id)}/acknowledge`, { method: "POST" })).data;
}

export async function fetchAlertRules(clusterId?: string) {
  const q = clusterId ? `?clusterId=${encodeURIComponent(clusterId)}` : "";
  const r = await api<AlertRule[]>(`/api/v1/alert-rules${q}`);
  return { rules: r.data ?? [], total: r.meta?.total ?? 0 };
}

export async function fetchAlertRuleMetrics() {
  return (await api<AlertRuleMetrics>("/api/v1/alert-rules/metrics")).data;
}

export async function createAlertRule(body: {
  name: string;
  message?: string;
  severity: AlertSeverity;
  metric: string;
  comparator: AlertComparator;
  threshold: number;
  enabled?: boolean;
  clusterId?: string;
}) {
  return (await api<AlertRule>("/api/v1/alert-rules", { method: "POST", body: JSON.stringify(body) })).data;
}

export async function updateAlertRule(
  id: string,
  body: Partial<{
    name: string;
    message: string;
    severity: AlertSeverity;
    metric: string;
    comparator: AlertComparator;
    threshold: number;
    enabled: boolean;
    clusterId: string;
    clearCluster: boolean;
  }>,
) {
  return (
    await api<AlertRule>(`/api/v1/alert-rules/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify(body),
    })
  ).data;
}

export async function deleteAlertRule(id: string) {
  await api<void>(`/api/v1/alert-rules/${encodeURIComponent(id)}`, { method: "DELETE" });
}

File: internal/domain/alerts.go
package domain

import (
	"slices"
	"time"
)

const (
	AlertSeverityInfo     = "info"
	AlertSeverityWarning  = "warning"
	AlertSeverityCritical = "critical"

	AlertStatusOpen   = "open"
	AlertStatusClosed = "closed"

	AlertComparatorGT  = "gt"
	AlertComparatorGTE = "gte"
	AlertComparatorLT  = "lt"
	AlertComparatorLTE = "lte"
)

var AlertSeverities = []string{AlertSeverityInfo, AlertSeverityWarning, AlertSeverityCritical}
var AlertComparators = []string{AlertComparatorGT, AlertComparatorGTE, AlertComparatorLT, AlertComparatorLTE}

// goalign:ignore
type AlertRule struct {
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
	Metric      string    `json:"metric"`
	Name        string    `json:"name"`
	Message     string    `json:"message"`
	Severity    string    `json:"severity"`
	ID          string    `json:"id"`
	Comparator  string    `json:"comparator"`
	CreatedBy   string    `json:"createdBy,omitempty"`
	AccountName string    `json:"accountName,omitempty"`
	ClusterID   string    `json:"clusterId,omitempty"`
	Threshold   float64   `json:"threshold"`
	Enabled     bool      `json:"enabled"`
}

type AlertRuleCreate struct {
	Enabled     *bool   `json:"enabled"`
	ClusterID   string  `json:"clusterId"`
	AccountName string  `json:"accountName"`
	Name        string  `json:"name"`
	Message     string  `json:"message"`
	Severity    string  `json:"severity"`
	Metric      string  `json:"metric"`
	Comparator  string  `json:"comparator"`
	Threshold   float64 `json:"threshold"`
}

// goalign:ignore
type AlertRuleUpdate struct {
	ClusterID    *string  `json:"clusterId"`
	AccountName  *string  `json:"accountName"`
	Name         *string  `json:"name"`
	Message      *string  `json:"message"`
	Severity     *string  `json:"severity"`
	Metric       *string  `json:"metric"`
	Comparator   *string  `json:"comparator"`
	Threshold    *float64 `json:"threshold"`
	Enabled      *bool    `json:"enabled"`
	ClearCluster bool     `json:"clearCluster"`
}

type Alert struct {
	FirstSeenAt    time.Time  `json:"firstSeenAt"`
	LastSeenAt     time.Time  `json:"lastSeenAt"`
	AcknowledgedAt *time.Time `json:"acknowledgedAt,omitempty"`
	ClosedAt       *time.Time `json:"closedAt,omitempty"`
	Status         string     `json:"status"`
	Severity       string     `json:"severity"`
	Metric         string     `json:"metric"`
	Message        string     `json:"message"`
	ID             string     `json:"id"`
	AccountName    string     `json:"accountName,omitempty"`
	ClusterID      string     `json:"clusterId"`
	RuleID         string     `json:"ruleId"`
	AcknowledgedBy string     `json:"acknowledgedBy,omitempty"`
	RuleName       string     `json:"ruleName,omitempty"`
	FiringValue    float64    `json:"firingValue"`
	Threshold      float64    `json:"threshold"`
}

type AlertFilter struct {
	Status     string
	ClusterID  string
	Severity   string
	ClusterIDs []string
	Limit      int
	Offset     int
}

type AlertOpenSummary struct {
	Alerts []Alert `json:"alerts"`
	Count  int     `json:"count"`
}

func ValidAlertSeverity(v string) bool {
	return slices.Contains(AlertSeverities, v)
}

func ValidAlertComparator(v string) bool {
	return slices.Contains(AlertComparators, v)
}

func ThresholdMet(comparator string, value, threshold float64) bool {
	switch comparator {
	case AlertComparatorGT:
		return value > threshold
	case AlertComparatorGTE:
		return value >= threshold
	case AlertComparatorLT:
		return value < threshold
	case AlertComparatorLTE:
		return value <= threshold
	default:
		return false
	}
}
